File: commonfamily/src/main/java/com/koenig/commonModel/finance/BankAccount.java
package com.koenig.commonModel.finance;

import com.koenig.FamilyConstants;
import com.koenig.commonModel.Item;
import com.koenig.commonModel.User;

import org.joda.time.DateTime;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by Thomas on 21.10.2015.
 */
public class BankAccount extends Item {
    private List<User> owners;
    private String bank;
    /**
     * Sorted list where the first one is the newest
     */
    private List<Balance> balances;

    public BankAccount(String id, String name, String bank, List<User> owners, List<Balance> balances) {
        super(id, name);
        this.owners = owners;
        this.name = name;
        this.bank = bank;
        this.balances = balances;
        sortBalances(balances);
    }

    public BankAccount(String id, String name, String bank, User owner, List<Balance> balances) {
        super(id, name);
        this.owners = new ArrayList<>(1);
        owners.add(owner);
        this.name = name;
        this.bank = bank;
        this.balances = balances;
        sortBalances(balances);
    }

    public BankAccount(String name, String bank, List<User> owners, List<Balance> balances) {
        super(name);
        this.owners = owners;
        this.name = name;
        this.bank = bank;
        this.balances = balances;
        sortBalances(balances);
    }

    public BankAccount(ByteBuffer buffer) {
        super(buffer);
        short size = buffer.getShort();
        this.owners = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            owners.add(new User(buffer));
        }

        bank = byteToString(buffer);
        size = buffer.getShort();
        balances = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            balances.add(new Balance(buffer));
        }
    }

    public static void sortBalances(List<Balance> balances) {
        // newest at top
        Collections.sort(balances, (lhs, rhs) -> rhs.getDate().compareTo(lhs.getDate()));
    }

    @Override
    public void writeBytes(ByteBuffer buffer) {
        super.writeBytes(buffer);
        writeList(owners, buffer);
        writeString(bank, buffer);
        writeList(balances, buffer);
    }

    public DateTime getDateTime() {
        if (balances.size() == 0) {
            return FamilyConstants.NO_DATE;
        }

        return balances.get(0).getDate();
    }

    public void addBalance(Balance balance) {
        balances.add(balance);
        sortBalances(balances);
    }

    @Override
    public String toString() {
        return "BankAccount{" +
                "id='" + id + '\'' +
                ", name='" + name + '\'' +
                ", owners=" + owners +
                ", bank='" + bank + '\'' +
                ", balances=" + balances +
                '}';
    }

    @Override
    public int getByteLength() {
        return super.getByteLength() + getListLength(owners) + getStringLength(bank) + getListLength(balances);
    }



    public List<Balance> getBalances() {
        return balances;
    }

    public void setBalances(List<Balance> balances) {
        this.balances = balances;
    }

    public List<User> getOwners() {
        return owners;
    }

    public void setOwners(List<User> owners) {
        this.owners = owners;
    }

    public String getBank() {
        return bank;
    }

    public void setBank(String bank) {
        this.bank = bank;
    }

    public int getBalance() {
        if (balances.size() == 0) return 0;

        return balances.get(0).getBalance();
    }

    public String toReadableString() {
        return getBank() + " - " + getName();
    }

    public void deleteBalance(Balance balance) {
        Balance removeBalance = null;
        for (Balance bal : balances) {
            if (bal.getDate().equals(balance.getDate())) {
                removeBalance = bal;
                break;
            }
        }

        balances.remove(removeBalance);
    }
}

File: commonfamily/src/main/java/com/koenig/commonModel/finance/BankAccount.kt
package com.koenig.commonModel.finance

import com.koenig.FamilyConstants
import com.koenig.commonModel.Byteable
import com.koenig.commonModel.Item
import com.koenig.commonModel.User
import org.joda.time.LocalDate
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by Thomas on 21.10.2015.
 */
class BankAccount : Item {
    var owners: MutableList<User>
    var bank: String
    /**
     * Sorted list where the first one is the newest
     */
    var balances: MutableList<Balance>

    // latest update of balances
    val dateTime: LocalDate
        get() = if (balances.size == 0) {
            FamilyConstants.BEGIN_LOCAL_DATE
        } else balances[0].day

    override val byteLength: Int
        get() = super.byteLength + Byteable.Companion.getListLength(owners) + Byteable.Companion.getStringLength(bank) + Byteable.Companion.getListLength(balances)

    val balance: Int
        get() = if (balances.size == 0) 0 else balances[0].balance

    constructor(id: String, name: String, bank: String, owners: MutableList<User>, balances: MutableList<Balance>) : super(id, name) {
        this.owners = owners
        this.name = name
        this.bank = bank
        this.balances = balances
        sortBalances(balances)
    }

    constructor(id: String, name: String, bank: String, owner: User, balances: MutableList<Balance>) : super(id, name) {
        this.owners = ArrayList(1)
        owners.add(owner)
        this.name = name
        this.bank = bank
        this.balances = balances
        sortBalances(balances)
    }

    constructor(name: String, bank: String, owners: MutableList<User>, balances: MutableList<Balance>) : super(name) {
        this.owners = owners
        this.name = name
        this.bank = bank
        this.balances = balances
        sortBalances(balances)
    }

    constructor(buffer: ByteBuffer) : super(buffer) {
        var size = buffer.short
        this.owners = ArrayList(size.toInt())
        for (i in 0 until size) {
            owners.add(User(buffer))
        }

        bank = Byteable.Companion.byteToString(buffer)
        size = buffer.short
        balances = ArrayList(size.toInt())
        for (i in 0 until size) {
            balances.add(Balance(buffer))
        }
    }

    override fun writeBytes(buffer: ByteBuffer) {
        super.writeBytes(buffer)
        Byteable.Companion.writeList(owners, buffer)
        Byteable.Companion.writeString(bank, buffer)
        Byteable.Companion.writeList(balances, buffer)
    }

    fun addBalance(balance: Balance) {
        balances.add(balance)
        sortBalances(balances)
    }

    override fun toString(): String {
        return "BankAccount{" +
                "id='" + id + '\''.toString() +
                ", name='" + name + '\''.toString() +
                ", owners=" + owners +
                ", bank='" + bank + '\''.toString() +
                ", balances=" + balances +
                '}'.toString()
    }


    fun toReadableString(): String {
        return bank + " - " + name
    }

    fun deleteBalance(balance: Balance) {
        var removeBalance: Balance? = null
        for (bal in balances) {
            if (bal.day == balance.day) {
                removeBalance = bal
                break
            }
        }

        balances.remove(removeBalance)
    }

    companion object {

        fun sortBalances(balances: List<Balance>) {
            // newest at top
            Collections.sort(balances) { lhs, rhs -> rhs.day.compareTo(lhs.day) }
        }
    }
}
